test.lua

```diff
@@ -781,6 +781,68 @@ do
     end
 end
 
+---- REPL tests ----
+print("Running tests for REPL completion...")
+local wrapRepl = function()
+    local replComplete
+    local replSend = coroutine.wrap(function()
+        local output = {}
+        fennel.repl({
+            readChunk = function()
+                local chunk = coroutine.yield(output)
+                output = {}
+                return chunk and chunk .. '\n' or nil
+            end,
+            onValues = function(xs)
+                table.insert(output, xs)
+            end,
+            registerCompleter = function(completer)
+                replComplete = function(input)
+                    local res = completer(input)
+                    table.sort(res)
+                    return res
+                end
+            end,
+            pp = function(x) return x end,
```

Comment:
Initially I wanted `pp` to just return the values since we're testing it from within lua anyway, and then use the already-implemented `deep_equal` function(s) to check the values. The use of `view` later was a workaround due to the current `deep_equal` having an infinite recursion bug when certain table shapes are passed to it (as I recall, `{{'foo'} {'foo'}}` caused it).

Longer-term I thought it would make sense to fix `deep_equal` so I just wrote that up in the meantime, but I can just as easily leave `pp` as-is, you're right.

Reply:
Looking at it in context (and considering the later discussion about just having the wrapped REPL return usable values), I think it still makes sense to replace `pp` just to avoid double-serialization, since the REPL can return multiple values.

Reply:
Hm; I can't reproduce the problem with `deep_equal`.

Reply:
Hrm, either I somehow fabricated this or it isn't happening with the tests as-is. Either way, it's cleaner with `deep_equal` so I'm going to push a version with the serialization stuff removed. Thanks for the sanity-check!